src/Data/Poly/Internal/Dense/Laurent.hs

```diff
@@ -0,0 +1,284 @@
+-- |
+-- Module:     Data.Poly.Internal.Laurent
+-- Copyright:  (c) 2019 Adjoint Inc
+-- Licence:    BSD3
+-- Maintainer: Adjoint Inc <info@adjoint.io>
+--
+-- Laurent polynomials of one variable.
+--
+{-# LANGUAGE DeriveAnyClass      #-}
+{-# LANGUAGE DeriveGeneric       #-}
+{-# LANGUAGE DerivingStrategies  #-}
+{-# LANGUAGE FlexibleContexts    #-}
+{-# LANGUAGE PatternSynonyms     #-}
+{-# LANGUAGE RankNTypes          #-}
+{-# LANGUAGE ScopedTypeVariables #-}
+{-# LANGUAGE ViewPatterns        #-}
+
+
+module Data.Poly.Internal.Dense.Laurent
+  ( Laurent(..)
+  , ULaurent
+  , VLaurent
+  , leading
+  , (^-)
+  -- * Num interface
```

Comment:
As mentioned in #23, it is enough to provide only a `Semiring`-based interface for Laurent polynomials.

Reply:
With only a Semiring-based interface for Laurent polynomials, it seems I can't define `subst` and `subst'`. Do you have any suggestions on this?

Reply:
You would not be able to define 

```haskell
subst :: (G.Vector v a, G.Vector w a, ...) => Laurent v a -> Laurent w a -> Laurent w a
```

anyways, because Laurent polynomials are not closed under composition: for example, you cannot substitute `Y = X+1` into `Y^(-1)`. But you can have 

```haskell
subst :: (G.Vector v a, G.Vector w a, Eq a, Field a) => Poly v a -> Laurent w a -> Laurent w a
```

with `Data.Field.Field` from `semirings` package.